pkg/clusteragent/autoscaling/cluster/model/node_pool_internal.go

```diff
@@ -0,0 +1,265 @@
+
+	// targetWeight is weight of the user-created NodePoolSpec
+	targetWeight *int32 `json:"target_weight"`
+}
+
+func NewNodePoolInternal(v *kubeAutoscaling.ClusterAutoscalingValues) NodePoolInternal {
+	return NodePoolInternal{
+		name:                     v.Name,
+		recommendedInstanceTypes: v.RecommendedInstanceTypes,
+		labels:                   convertLabels(v.Labels),
+		taints:                   convertTaints(v.Taints),
+		targetName:               v.TargetName,
+		targetHash:               v.TargetHash,
+		targetWeight:             v.TargetWeight,
+	}
+}
+
+func ConvertToKarpenterNodePool(n NodePoolInternal, nodeClassName string) *karpenterv1.NodePool {
+	return &karpenterv1.NodePool{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "NodePool",
+			APIVersion: "karpenter.sh/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   n.name,
+			Labels: map[string]string{DatadogCreatedLabelKey: "true"},
+		},
+		Spec: buildNodePoolSpec(n, nodeClassName),
+	}
+
+}
+
+// Getters
+
+// Name returns the name of the NodePoolInternal
+func (n *NodePoolInternal) Name() string {
+	return n.name
+}
+
+// RecommendedInstanceTypes returns the recommendedInstanceTypes of the NodePoolInternal
+func (n *NodePoolInternal) RecommendedInstanceTypes() []string {
+	return n.recommendedInstanceTypes
+}
+
+// Labels returns the labels of the NodePoolInternal
+func (n *NodePoolInternal) Labels() map[string]string {
+	return n.labels
+}
+
+// Taints returns the taints of the NodePoolInternal
+func (n *NodePoolInternal) Taints() []corev1.Taint {
+	return n.taints
+}
+
+// TargetName returns the targetName of the NodePoolInternal
+func (n *NodePoolInternal) TargetName() string {
+	return n.targetName
+}
+
+// TargetHash returns the targetHash of the NodePoolInternal
+func (n *NodePoolInternal) TargetHash() string {
+	return n.targetHash
+}
+
+// TargetWeight returns the targetWeight of the NodePoolInternal
+func (n *NodePoolInternal) TargetWeight() int32 {
+	return *n.targetWeight
+}
+
+func convertLabels(input []*kubeAutoscaling.DomainLabels) map[string]string {
+	output := make(map[string]string)
+	for _, elem := range input {
+		output[elem.Key] = elem.Value
+	}
+
+	return output
+}
+
+func convertTaints(input []*kubeAutoscaling.Taints) []corev1.Taint {
+	output := []corev1.Taint{}
+	for _, elem := range input {
+		output = append(output, corev1.Taint{
+			Key:    elem.Key,
+			Value:  elem.Value,
+			Effect: corev1.TaintEffect(elem.Effect),
+		})
+	}
+
+	return output
+}
+
+var deprecatedLabels = sets.New[string]("beta.kubernetes.io/arch", "beta.kubernetes.io/os")
+
+// buildNodePoolSpec is used for creating new NodePools
+func buildNodePoolSpec(n NodePoolInternal, nodeClassName string) karpenterv1.NodePoolSpec {
+
+	wellKnownLabels := karpenterv1.WellKnownLabels
+
+	metadataLabels := map[string]string{}
+
+	// Convert domain labels into requirements
+	reqs := []karpenterv1.NodeSelectorRequirementWithMinValues{}
```

Comment:
In order to do fewer allocations:
```suggestion
	reqs := make([]karpenterv1.NodeSelectorRequirementWithMinValues, 0, len(n.labels)+1)
```